### ISSUE TITLE
Incompatibility with strict CSPs

### ISSUE THREAD

**Issue by arnemolland** [emilkowalski/sonner#449]
### Describe the feature / bug 📝:

Sonner does not work with strict CSPs as there's inline styles.

### Steps to reproduce the bug 🔁:

1. Use any CSP with the `style-src` directive set to anything other than `unsafe-inline`.

I'm looking into some fixes, but in essence all styles have to be defined in stylesheets for other CSPs to work. If it's impractical to apply a fix, there's always the fork and modify path but having support baked in would be nice.

**Comment by arnemolland** [comment]
Worth mentioning if someone else stumbles across this; I'm currently (manually) creating hashes of the computed inline styles and using `unsafe-hashes` to allow them.

**Comment by BaDo2001** [comment]
+1

**Comment by louis-foucart** [comment]
+1

**Comment by WoetDev** [comment]
+1

**Comment by AkshayCloudAnalogy** [comment]
+1

**Comment by willrp** [comment]
+1

**Comment by Pruxis** [comment]
+1

**Comment by emilkowalski** [comment]
The solution here would be moving into a stylesheet instead of injecting the styles into the head, am I understanding it correctly?

**Comment by hugomrdias** [comment]
yes, dont bundle css you are already publishing the style.css file we can just import it the way makes more sense on our side.

**Comment by pleunv** [comment]
Personally I avoid libraries that do any sort of script/src injection _because_ of the CSP complications it introduces and instead prefer libraries that force the consumer to manually import the necessary assets (i.e. css) so I can let the bundler handle it with their centralized CSP config.

I do agree that Vite's (and webpack's, and probably any bundler's) CSP approach feels kind of insecure, however, I don't think there's really any viable alternative at the momen. If you are employing a secure CSP then it shouldn't matter that scripts can parse the meta tag because the CSP will prevent those scripts from running in the first place. In addition, CSP nonces are non-deterministic and refreshed on every request, making them impossible to be reused.

**Comment by Gunnbra** [comment]
Hello, is there any movement on this?

**Comment by emilkowalski** [comment]
We've looked into it with @huozhi, but decided not to support it. Requiring users to import the styles manually is bad DX and there doesn't seem to be a good solution that would support both.

Feel free to fork this repo and import the styles manually instead of injecting them.

**Comment by Gunnbra** [comment]
Another solution I found is copying the CSS from the Sonner Github and putting it into your own project

**Comment by TmLev** [comment]
> We've looked into it with @huozhi, but decided not to support it. Requiring users to import the styles manually is bad DX and there doesn't seem to be a good solution that would support both.

Bad DX is definitely not desirable, but with strict CSP policies the library doesn't work.
I'd say adding one import line is preferable over not being able to use Sonner at all.

**Comment by ryanstaniforth** [comment]
> Requiring users to import the styles manually is bad DX and there doesn't seem to be a good solution that would support both.

Having to import styles with a one line change is not bad DX. Bad DX is having to hack around a library that doesn't play well with browser standards.

**Comment by pleunv** [comment]
I don't think there's a need for passive aggressiveness, Emil is maintaining this in his spare time and is in no way obliged to meet everyone's demands. Nobody's forcing you to use this particular library and you can also fork it.

And to make this a slightly more constructive post, for those in the same CSP boat as myself. When looking at alternatives I've found that many of the toast libraries have exactly the same (embedded) CSP setup as sonner. In that list:
- https://github.com/iamhosseindhv/notistack
- https://github.com/fkhadra/react-toastify
- https://github.com/timolins/react-hot-toast
- https://github.com/pheralb/toast

Of these 4, `react-hot-toast` provides a dedicated import path that _does not_ do an automatic embedding of the `css` file, see https://github.com/fkhadra/react-toastify/issues/1215. `notistack` and `react-toastify` are built on top of https://github.com/cristianbote/goober, which has a built-in CSP workaround that can be used by both libraries to make them work without an `unsafe-inline` CSP directive by basically predefining a `<style nonce="…" id="_goober"></style>` element (see https://github.com/cristianbote/goober/issues/471 and https://github.com/cristianbote/goober/issues/471#issuecomment-1981327546). https://github.com/pheralb/toast sadly has a very similar setup as sonner and cannot be worked around. Personally I'm currently sticking with `react-hot-toast` until `react-aria-components`'s [Toast](https://react-spectrum.adobe.com/react-aria/Toast.html) components comes out of alpha.

**Comment by YagoGG** [comment]
FYI, I was able to work around it by explicitly importing sonner's CSS:

```js
import 'sonner/dist/styles.css';
```

In my case (using Next.js with the App Router), just had to add it to the same `layout.tsx` where I had put the `<Toast>` component.

If your build system is set up correctly, it should inject the CSS with the nonce (as the rest of the CSS in your app). Hope that helps!
